pkg/operator/controllers/muo/muo_controller.go

```diff
@@ -0,0 +1,161 @@
+)
+
+const (
+	enabled  string = "rh.srep.muo.enabled"
+	managed  string = "rh.srep.muo.managed"
+	pullSpec string = "rh.srep.muo.deploy.pullspec"
+)
+
+type Reconciler struct {
+	arocli        aroclient.Interface
+	kubernetescli kubernetes.Interface
+	dh            dynamichelper.Interface
+
+	readinessPollTime time.Duration
+	readinessTimeout  time.Duration
+}
+
+func NewReconciler(log *logrus.Entry, arocli aroclient.Interface, kubernetescli kubernetes.Interface, dh dynamichelper.Interface) *Reconciler {
+	return &Reconciler{
+		arocli:        arocli,
+		kubernetescli: kubernetescli,
+		dh:            dh,
+
+		readinessPollTime: 10 * time.Second,
+		readinessTimeout:  5 * time.Minute,
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
+	instance, err := r.arocli.AroV1alpha1().Clusters().Get(ctx, arov1alpha1.SingletonClusterName, metav1.GetOptions{})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if !instance.Spec.OperatorFlags.GetSimpleBoolean(enabled) {
+		// controller is disabled
+		condition := &operatorv1.OperatorCondition{
+			Type:    arov1alpha1.ManagedUpgradeOperatorStatus,
+			Status:  operatorv1.ConditionUnknown,
+			Message: "Controller is disabled",
+			Reason:  "ControllerDisabled",
+		}
+		err = conditions.SetCondition(ctx, r.arocli, condition, "master")
+		return reconcile.Result{}, err
+	}
+
+	managed := instance.Spec.OperatorFlags.GetWithDefault(managed, "")
+	deployer := NewMUODeployer(instance, r.kubernetescli, r.dh)
+
+	var condition *operatorv1.OperatorCondition
+
+	// If enabled and managed=true, install MUO
+	// If enabled and managed=false, remove the MUO deployment
+	// If enabled and managed is missing, do nothing
+	if strings.EqualFold(managed, "true") {
+		err = deployer.CreateOrUpdate(ctx)
+		if err != nil {
+			condition = &operatorv1.OperatorCondition{
+				Type:    arov1alpha1.ManagedUpgradeOperatorStatus,
+				Status:  operatorv1.ConditionFalse,
+				Message: err.Error(),
+				Reason:  "DeploymentFailed",
+			}
+			err = conditions.SetCondition(ctx, r.arocli, condition, "master")
+			return reconcile.Result{}, err
+		}
+
+		// Check that MUO has become ready, wait up to readinessTimeout (default 5min)
+		timeoutCtx, cancel := context.WithTimeout(ctx, r.readinessTimeout)
+		defer cancel()
+
+		err := wait.PollImmediateUntil(r.readinessPollTime, func() (bool, error) {
+			return deployer.IsReady(ctx)
+		}, timeoutCtx.Done())
+
+		if err != nil {
+			condition = &operatorv1.OperatorCondition{
+				Type:    arov1alpha1.ManagedUpgradeOperatorStatus,
+				Status:  operatorv1.ConditionFalse,
+				Message: "Deployment did not become ready",
+				Reason:  "DeploymentFailed",
+			}
+		} else {
+			condition = &operatorv1.OperatorCondition{
+				Type:    arov1alpha1.ManagedUpgradeOperatorStatus,
+				Status:  operatorv1.ConditionTrue,
+				Message: "Deployment is ready",
+				Reason:  "DeploymentSucceeded",
+			}
+		}
```

Comment:
Do we really care about exposing these conditions?

---

What do you think if we do the following?

In the controller
* Wait for the deployment to become ready after `deployer.CreateOrUpdate`.
  * If we error waiting (currently line 98) for deployment - return this error to requeue.
  * If ready - all good
 * Do not wait after `deployer.Remove`

This way we make sure that we do not try to create or update multiple times while containers are still creating.

In the monitor:
* Implement something like [ARO operator heartbeat](https://github.com/Azure/ARO-RP/blob/efe990b0a39f7f4889d452da99f38488e4c6c1e1/pkg/monitor/cluster/arooperatorheartbeat.go), but for MUO.
* Potentially create an alert on it

This way we monitor clusters with enabled and managed MUO operator and do not monitor everything else... And we don't have to deal with a mess of conditions.

If after `deployer.CreateOrUpdate` dpeloyment doesn't become ready - we should pick this up with the monitor. Should we care about any other cases?

Reply:
Fair -- I am not super sure how we want to monitor all of this yet, since there's PrometheusRule stuff from OSD that we can't use that triggers Prometheus alerts. I've removed it for now, we can implement a better version before full GA.